src/thread_pool_async.ml

```diff
@@ -2,83 +2,113 @@ open Base
 open Async_kernel
 open Async_unix
 
-type 'state thread = In_thread.Helper_thread.t * 'state
+type 'state worker = {
+  thread : In_thread.Helper_thread.t;
+  state : 'state;
+}
+
+exception Invalid_thread_count of int
+
+exception Pool_already_destroyed
+
+exception State_manipulation_error of exn
 
 type 'state t = {
-  pool : 'state thread Pipe.Reader.t * 'state thread Pipe.Writer.t;
+  reader : 'state worker Pipe.Reader.t;
+  writer : 'state worker Pipe.Writer.t;
   create : unit -> 'state;
   destroy : 'state -> unit;
-  threads : int;
+  thread_count : int;
 }
 
-let init ~name ~threads ~create ~destroy =
-  let create_thread ~name ~create =
-    let%bind thread = In_thread.Helper_thread.create ~name () in
-    let%bind state = In_thread.run ~thread create in
-    return (thread, state)
+let try_in_thread ~thread f = try_with (fun () -> In_thread.run ~thread f)
+
+let destroy_pool {reader; writer; thread_count; destroy; _} =
+  let%bind states_destroyed =
+    List.init thread_count ~f:ignore
+    |> Deferred.List.map ~how:`Parallel ~f:(fun () ->
+           match%bind Pipe.read reader with
+           | `Eof -> return @@ Ok ()
+           | `Ok {thread; state} -> try_in_thread ~thread (fun () -> destroy state))
+    |> Deferred.map ~f:Result.all_unit
   in
-  (* No strict ordering on the heap. All items are equal *)
-  let reader, writer = Pipe.create () in
-  let elements = List.init threads ~f:ignore in
-  let add_to_pool () =
-    create_thread ~name ~create >>| Pipe.write_without_pushback writer
+  Pipe.close writer;
+  match states_destroyed with
+  | Ok () -> return ()
+  | Error exn -> raise @@ State_manipulation_error exn
+
+let recreate_state ({create; destroy; _} as pool) {thread; state} =
+  let destroy_pool_and_raise exn =
+    let%bind () = destroy_pool pool in
+    raise @@ State_manipulation_error exn
   in
-  let%bind () = Deferred.List.iter elements ~how:`Parallel ~f:add_to_pool in
-  return {pool = reader, writer; create; destroy; threads}
+  match%bind try_in_thread ~thread (fun () -> destroy state) with
+  | Ok () -> (
+      match%bind try_in_thread ~thread create with
+      | Ok state -> return {thread; state}
+      | Error exn -> destroy_pool_and_raise exn)
+  | Error exn -> destroy_pool_and_raise exn
 
-type 'result computation =
-  [ `Ok of 'result
-  | `Attempt_retry ]
+let init ~name ~threads:thread_count ~create ~destroy =
+  if thread_count <= 0
```

Comment:
I would use `match` here because I find the formatting and precedence of `if` absolutely confusing but that might be just me.

Reply:
Yeah, I don't know. Feel free to change it if you like.